diff --git a/drivers/iio/adc/ad9361.c b/drivers/iio/adc/ad9361.c
index 8583e7fe..ebf8f756 100644
--- a/drivers/iio/adc/ad9361.c
+++ b/drivers/iio/adc/ad9361.c
@@ -1234,7 +1234,7 @@ static int ad9361_load_mixer_gm_subtable(struct ad9361_rf_phy *phy)
 	return 0;
 }
 
-static int ad9361_set_tx_atten(struct ad9361_rf_phy *phy, u32 atten_mdb,
+int ad9361_set_tx_atten(struct ad9361_rf_phy *phy, u32 atten_mdb,
 			       bool tx1, bool tx2, bool immed)
 {
 	u8 buf[2];
@@ -1266,8 +1266,9 @@ static int ad9361_set_tx_atten(struct ad9361_rf_phy *phy, u32 atten_mdb,
 
 	return ret;
 }
+EXPORT_SYMBOL(ad9361_set_tx_atten);
 
-static int ad9361_get_tx_atten(struct ad9361_rf_phy *phy, u32 tx_num)
+int ad9361_get_tx_atten(struct ad9361_rf_phy *phy, u32 tx_num)
 {
 	u8 buf[2];
 	int ret = 0;
@@ -1285,6 +1286,7 @@ static int ad9361_get_tx_atten(struct ad9361_rf_phy *phy, u32 tx_num)
 
 	return code;
 }
+EXPORT_SYMBOL(ad9361_get_tx_atten);
 
 int ad9361_tx_mute(struct ad9361_rf_phy *phy, u32 state)
 {
@@ -2829,7 +2831,7 @@ static int ad9361_tx_quad_calib(struct ad9361_rf_phy *phy,
 	else
 		decim = 3;
 
-	if (clkrf == (2 * clktf)) {
+	if ((clkrf / 2) == clktf) {
 		__rx_phase = 0x0E;
 		switch (txnco_word) {
 		case 0:
@@ -2864,7 +2866,7 @@ static int ad9361_tx_quad_calib(struct ad9361_rf_phy *phy,
 				__rx_phase = 0x1A;
 			break;
 		}
-	} else if (clktf == (2 * clkrf)) {
+	} else if ((clktf / 2) == clkrf) {
 		__rx_phase = -2;
 		switch (txnco_word) {
 		case 0:
@@ -2877,11 +2879,11 @@ static int ad9361_tx_quad_calib(struct ad9361_rf_phy *phy,
 			rxnco_word = 3;
 			break;
 		}
-	} else if (clktf == (4 * clkrf)) {
+	} else if ((clktf / 4) == clkrf) {
 		__rx_phase = -2;
 		txnco_word = 0;
 		rxnco_word = 3;
-	} else if (clkrf == (4 * clktf)) {
+	} else if ((clkrf / 4) == clktf) {
 		__rx_phase = -2;
 		txnco_word = 3;
 		rxnco_word = 0;
@@ -3449,6 +3451,8 @@ static int ad9361_gc_setup(struct ad9361_rf_phy *phy, struct gain_control *ctrl)
 			  AGCLL_MAX_INCREASE(~0),  reg);
 
 	/* Fast AGC - Peak Detectors and Final Settling */
+  ad9361_spi_writef(spi, REG_AGC_LOCK_LEVEL, ENABLE_DIG_SAT_OVRG, ctrl->f_agc_dig_sat_ovrg_en);
+
 	reg = ctrl->f_agc_lpf_final_settling_steps;
 	reg = clamp_t(u32, reg, 0U, 3U);
 	ad9361_spi_writef(spi, REG_FAST_ENERGY_LOST_THRESH,
@@ -3744,7 +3748,7 @@ static int ad9361_get_auxadc(struct ad9361_rf_phy *phy)
   // Setup Control Outs
   //************************************************************
 
-static int ad9361_ctrl_outs_setup(struct ad9361_rf_phy *phy,
+int ad9361_ctrl_outs_setup(struct ad9361_rf_phy *phy,
 				  struct ctrl_outs_control *ctrl)
 {
 	struct spi_device *spi = phy->spi;
@@ -3754,6 +3758,7 @@ static int ad9361_ctrl_outs_setup(struct ad9361_rf_phy *phy,
 	ad9361_spi_write(spi, REG_CTRL_OUTPUT_POINTER, ctrl->index); // Ctrl Out index
 	return ad9361_spi_write(spi, REG_CTRL_OUTPUT_ENABLE, ctrl->en_mask); // Ctrl Out [7:0] output enable
 }
+EXPORT_SYMBOL(ad9361_ctrl_outs_setup);
   //************************************************************
   // Setup GPO
   //************************************************************
@@ -4572,7 +4577,7 @@ static int ad9361_fastlock_store(struct ad9361_rf_phy *phy, bool tx, u32 profile
 	/* Wide BW option: N = 1
 	 * Set init and steady state values to the same - let user space handle it
 	 */
-	val[6] = (x << 3) | y;
+	val[6] = (x << 6) | y;
 	val[7] = y;
 
 	x = ad9361_spi_readf(spi, REG_RX_LOOP_FILTER_3 + offs, LOOP_FILTER_R3(~0));
@@ -4939,6 +4944,11 @@ static unsigned long ad9361_ref_div_sel(unsigned long refin_Hz, unsigned long ma
 		return 0;
 }
 
+static void ad9361_clk_disable(void *clk)
+{
+	clk_disable_unprepare(clk);
+}
+
 static int ad9361_setup(struct ad9361_rf_phy *phy)
 {
 	struct ad9361_rf_phy_state *st = phy->state;
@@ -5005,6 +5015,10 @@ static int ad9361_setup(struct ad9361_rf_phy *phy)
 	if (ret < 0)
 		return ret;
 
+	ret = devm_add_action_or_reset(dev, ad9361_clk_disable, phy->clk_refin);
+	if (ret)
+		return ret;
+
 	ret = clk_set_rate(phy->clks[BB_REFCLK], ref_freq);
 	if (ret < 0) {
 		dev_err(dev, "Failed to set BB ref clock rate (%d)\n",
@@ -5104,6 +5118,10 @@ static int ad9361_setup(struct ad9361_rf_phy *phy)
 	if (ret < 0)
 		return ret;
 
+	ret = devm_add_action_or_reset(dev, ad9361_clk_disable, phy->clks[RX_RFPLL]);
+	if (ret)
+		return ret;
+
 	/* Skip quad cal here we do it later again */
 	st->last_tx_quad_cal_freq = pd->tx_synth_freq;
 	ret = clk_set_rate(phy->clks[TX_RFPLL], ad9361_to_clk(pd->tx_synth_freq));
@@ -5117,6 +5135,10 @@ static int ad9361_setup(struct ad9361_rf_phy *phy)
 	if (ret < 0)
 		return ret;
 
+	ret = devm_add_action_or_reset(dev, ad9361_clk_disable, phy->clks[TX_RFPLL]);
+	if (ret)
+		return ret;
+
 	clk_set_parent(phy->clks[RX_RFPLL],
 			pd->use_ext_rx_lo ? phy->clk_ext_lo_rx :
 			phy->clks[RX_RFPLL_INT]);
@@ -5235,7 +5257,7 @@ static int ad9361_setup(struct ad9361_rf_phy *phy)
 
 }
 
-static int ad9361_do_calib_run(struct ad9361_rf_phy *phy, u32 cal, int arg)
+int ad9361_do_calib_run(struct ad9361_rf_phy *phy, u32 cal, int arg)
 {
 	struct ad9361_rf_phy_state *st = phy->state;
 	int ret;
@@ -5268,6 +5290,7 @@ static int ad9361_do_calib_run(struct ad9361_rf_phy *phy, u32 cal, int arg)
 
 	return ret;
 }
+EXPORT_SYMBOL(ad9361_do_calib_run);
 
 static int ad9361_update_rf_bandwidth(struct ad9361_rf_phy *phy,
 				     u32 rf_rx_bw, u32 rf_tx_bw)
@@ -5740,7 +5763,10 @@ static inline int ad9361_set_muldiv(struct refclk_scale *priv, u32 mul, u32 div)
 static int ad9361_get_clk_scaler(struct clk_hw *hw)
 {
 	struct refclk_scale *clk_priv = to_clk_priv(hw);
+	struct ad9361_rf_phy *phy = clk_priv->phy;
+	struct ad9361_phy_platform_data *pd = phy->pdata;
 	struct spi_device *spi = clk_priv->spi;
+	const bool lvds_mode = pd->port_ctrl.pp_conf[2] & LVDS_MODE;
 	u32 tmp, tmp1;
 
 	switch (clk_priv->source) {
@@ -5805,6 +5831,8 @@ static int ad9361_get_clk_scaler(struct clk_hw *hw)
 			tmp = (1 << (tmp - 1));
 
 		return ad9361_set_muldiv(clk_priv, 1, tmp);
+	case PL_INTF_CLK:
+		return ad9361_set_muldiv(clk_priv, (ad9361_uses_rx2tx2(phy) + 1) << lvds_mode, 1);
 	default:
 		return -EINVAL;
 	}
@@ -5973,6 +6001,8 @@ static int ad9361_set_clk_scaler(struct clk_hw *hw, bool set)
 			return ad9361_spi_writef(spi, REG_TX_ENABLE_FILTER_CTRL,
 					TX_FIR_ENABLE_INTERPOLATION(~0), tmp);
 		break;
+	case PL_INTF_CLK:
+		break;
 	default:
 		return -EINVAL;
 	}
@@ -6049,6 +6079,11 @@ static const struct clk_ops refclk_scale_ops = {
 	.recalc_rate = ad9361_clk_factor_recalc_rate,
 };
 
+static const struct clk_ops pl_intf_clk_ops = {
+	.round_rate = ad9361_clk_factor_round_rate,
+	.recalc_rate = ad9361_clk_factor_recalc_rate,
+};
+
 /*
  * BBPLL
  */
@@ -6570,6 +6605,9 @@ static int ad9361_clk_register(struct ad9361_rf_phy *phy,
 	case TX_RFPLL:
 		init.ops = &rfpll_clk_ops;
 		break;
+	case PL_INTF_CLK:
+		init.ops = &pl_intf_clk_ops;
+		break;
 	default:
 		init.ops = &refclk_scale_ops;
 	}
@@ -6598,6 +6636,20 @@ static int ad9361_clks_resync(struct ad9361_rf_phy *phy)
 	return 0;
 }
 
+static void ad9361_unregister_rx_notifier(void *data)
+{
+	struct ad9361_rf_phy *phy = data;
+
+	clk_notifier_unregister(phy->clks[RX_RFPLL], &phy->clk_nb_rx);
+}
+
+static void ad9361_unregister_tx_notifier(void *data)
+{
+	struct ad9361_rf_phy *phy = data;
+
+	clk_notifier_unregister(phy->clks[TX_RFPLL], &phy->clk_nb_tx);
+}
+
 static int register_clocks(struct ad9361_rf_phy *phy)
 {
 	const char *parent_name;
@@ -6686,6 +6738,9 @@ static int register_clocks(struct ad9361_rf_phy *phy)
 	ad9361_clk_register(phy, "-tx_sampl_clk", "-clktf_clk", NULL,
 		flags | CLK_IGNORE_UNUSED, TX_SAMPL_CLK);
 
+	ad9361_clk_register(phy, "-pl_intf_clk", "-tx_sampl_clk", NULL,
+		flags | CLK_IGNORE_UNUSED, PL_INTF_CLK);
+
 	ad9361_clk_register(phy, "-rx_rfpll_int", "-rx_refclk", NULL,
 		flags | CLK_IGNORE_UNUSED, RX_RFPLL_INT);
 
@@ -6705,11 +6760,19 @@ static int register_clocks(struct ad9361_rf_phy *phy)
 	if (ret < 0)
 		return ret;
 
+	ret = devm_add_action_or_reset(&phy->spi->dev, ad9361_unregister_rx_notifier, phy);
+	if (ret)
+		return ret;
+
 	phy->clk_nb_tx.notifier_call = ad9361_tx_rfpll_rate_change;
 	ret = clk_notifier_register(phy->clks[TX_RFPLL], &phy->clk_nb_tx);
 	if (ret < 0)
 		return ret;
 
+	ret = devm_add_action_or_reset(&phy->spi->dev, ad9361_unregister_tx_notifier, phy);
+	if (ret)
+		return ret;
+
 	return 0;
 }
 
@@ -6757,7 +6820,7 @@ static ssize_t ad9361_phy_store(struct device *dev,
 		this_attr->address != AD9361_ENSM_MODE)
 		return -EINVAL;
 
-	mutex_lock(&indio_dev->mlock);
+	mutex_lock(&phy->lock);
 
 	switch ((u32)this_attr->address) {
 	case AD9361_RF_RX_BANDWIDTH:
@@ -7018,7 +7081,7 @@ static ssize_t ad9361_phy_store(struct device *dev,
 		ret = -EINVAL;
 	}
 
-	mutex_unlock(&indio_dev->mlock);
+	mutex_unlock(&phy->lock);
 
 	return ret ? ret : len;
 }
@@ -7035,7 +7098,7 @@ static ssize_t ad9361_phy_show(struct device *dev,
 	unsigned long clk[6];
 	u64 delta;
 
-	mutex_lock(&indio_dev->mlock);
+	mutex_lock(&phy->lock);
 	switch ((u32)this_attr->address) {
 	case AD9361_RF_RX_BANDWIDTH:
 		ret = sprintf(buf, "%u\n", st->current_rx_bw_Hz);
@@ -7153,7 +7216,7 @@ static ssize_t ad9361_phy_show(struct device *dev,
 	default:
 		ret = -EINVAL;
 	}
-	mutex_unlock(&indio_dev->mlock);
+	mutex_unlock(&phy->lock);
 
 	return ret;
 }
@@ -7324,14 +7387,14 @@ static int ad9361_phy_reg_access(struct iio_dev *indio_dev,
 	struct ad9361_rf_phy *phy = iio_priv(indio_dev);
 	int ret;
 
-	mutex_lock(&indio_dev->mlock);
+	mutex_lock(&phy->lock);
 	if (readval == NULL) {
 		ret = ad9361_spi_write(phy->spi, reg, writeval);
 	} else {
 		*readval =  ad9361_spi_read(phy->spi, reg);
 		ret = 0;
 	}
-	mutex_unlock(&indio_dev->mlock);
+	mutex_unlock(&phy->lock);
 
 	return ret;
 }
@@ -7384,7 +7447,7 @@ static ssize_t ad9361_phy_lo_write(struct iio_dev *indio_dev,
 			break;
 	}
 
-	mutex_lock(&indio_dev->mlock);
+	mutex_lock(&phy->lock);
 	switch (private) {
 	case LOEXT_FREQ:
 		switch (chan->channel) {
@@ -7480,7 +7543,7 @@ static ssize_t ad9361_phy_lo_write(struct iio_dev *indio_dev,
 
 		break;
 	}
-	mutex_unlock(&indio_dev->mlock);
+	mutex_unlock(&phy->lock);
 
 	return ret ? ret : len;
 }
@@ -7496,7 +7559,7 @@ static ssize_t ad9361_phy_lo_read(struct iio_dev *indio_dev,
 	size_t len;
 	int ret = 0;
 
-	mutex_lock(&indio_dev->mlock);
+	mutex_lock(&phy->lock);
 	switch (private) {
 	case LOEXT_FREQ:
 		val = ad9361_from_clk(clk_get_rate(phy->clks[chan->channel ?
@@ -7513,7 +7576,7 @@ static ssize_t ad9361_phy_lo_read(struct iio_dev *indio_dev,
 			len += sprintf(buf + len, "%u%c", faslock_vals[i],
 				       i == 15 ? '\n' : ',');
 
-		mutex_unlock(&indio_dev->mlock);
+		mutex_unlock(&phy->lock);
 		return len;
 		}
 	case LOEXT_RECALL:
@@ -7553,14 +7616,14 @@ static ssize_t ad9361_phy_lo_read(struct iio_dev *indio_dev,
 		}
 
 		len = sprintf(buf, "[%llu 1 %llu]\n", min, max);
-		mutex_unlock(&indio_dev->mlock);
+		mutex_unlock(&phy->lock);
 		return len;
 		}
 	default:
 		ret = 0;
 
 	}
-	mutex_unlock(&indio_dev->mlock);
+	mutex_unlock(&phy->lock);
 
 	return ret < 0 ? ret : sprintf(buf, "%llu\n", val);
 }
@@ -7690,7 +7753,7 @@ static ssize_t ad9361_phy_rx_write(struct iio_dev *indio_dev,
 				    const struct iio_chan_spec *chan,
 				    const char *buf, size_t len)
 {
-//	struct ad9361_rf_phy *phy = iio_priv(indio_dev);
+	struct ad9361_rf_phy *phy = iio_priv(indio_dev);
 	u64 readin;
 	int ret = 0;
 
@@ -7698,7 +7761,7 @@ static ssize_t ad9361_phy_rx_write(struct iio_dev *indio_dev,
 	if (ret)
 		return ret;
 
-	mutex_lock(&indio_dev->mlock);
+	mutex_lock(&phy->lock);
 	switch (chan->channel) {
 	case 0:
 
@@ -7712,7 +7775,7 @@ static ssize_t ad9361_phy_rx_write(struct iio_dev *indio_dev,
 		ret = -EINVAL;
 		ret = 0;
 	}
-	mutex_unlock(&indio_dev->mlock);
+	mutex_unlock(&phy->lock);
 
 	return ret ? ret : len;
 }
@@ -7727,14 +7790,14 @@ static ssize_t ad9361_phy_rx_read(struct iio_dev *indio_dev,
 	int val;
 	int ret = 0;
 
-	mutex_lock(&indio_dev->mlock);
+	mutex_lock(&phy->lock);
 
 	rssi.ant = ad9361_1rx1tx_channel_map(phy, false, chan->channel + 1);
 	rssi.duration = 1;
 	ret = ad9361_read_rssi(phy, &rssi);
 	val = rssi.symbol;
 
-	mutex_unlock(&indio_dev->mlock);
+	mutex_unlock(&phy->lock);
 
 	return ret < 0 ? ret : sprintf(buf, "%u.%02u dB\n",
 			val / rssi.multiplier, val % rssi.multiplier);
@@ -7757,7 +7820,7 @@ static ssize_t ad9361_phy_tx_read(struct iio_dev *indio_dev,
 	u32 val;
 	int ret;
 
-	mutex_lock(&indio_dev->mlock);
+	mutex_lock(&phy->lock);
 	ret = ad9361_spi_readm(phy->spi, REG_TX_RSSI_LSB,
 			reg_val_buf, ARRAY_SIZE(reg_val_buf));
 
@@ -7771,7 +7834,7 @@ static ssize_t ad9361_phy_tx_read(struct iio_dev *indio_dev,
 	default:
 		ret = -EINVAL;
 	}
-	mutex_unlock(&indio_dev->mlock);
+	mutex_unlock(&phy->lock);
 
 	val *= RSSI_RESOLUTION;
 
@@ -7814,7 +7877,7 @@ static int ad9361_phy_read_raw(struct iio_dev *indio_dev,
 	struct ad9361_rf_phy *phy = iio_priv(indio_dev);
 	int ret;
 
-	mutex_lock(&indio_dev->mlock);
+	mutex_lock(&phy->lock);
 	switch (m) {
 	case IIO_CHAN_INFO_HARDWAREGAIN:
 		if (chan->output) {
@@ -7892,7 +7955,7 @@ static int ad9361_phy_read_raw(struct iio_dev *indio_dev,
 	}
 
 out_unlock:
-	mutex_unlock(&indio_dev->mlock);
+	mutex_unlock(&phy->lock);
 
 	return ret;
 };
@@ -7911,7 +7974,7 @@ static int ad9361_phy_write_raw(struct iio_dev *indio_dev,
 	if (st->curr_ensm_state == ENSM_STATE_SLEEP)
 		return -EINVAL;
 
-	mutex_lock(&indio_dev->mlock);
+	mutex_lock(&phy->lock);
 	switch (mask) {
 	case IIO_CHAN_INFO_HARDWAREGAIN:
 		if (chan->output) {
@@ -7966,7 +8029,7 @@ static int ad9361_phy_write_raw(struct iio_dev *indio_dev,
 		ret = -EINVAL;
 	}
 out:
-	mutex_unlock(&indio_dev->mlock);
+	mutex_unlock(&phy->lock);
 
 	return ret;
 }
@@ -8162,10 +8225,10 @@ static ssize_t ad9361_debugfs_read(struct file *file, char __user *userbuf,
 
 	} else if (entry->cmd == DBGFS_RXGAIN_1 || entry->cmd == DBGFS_RXGAIN_2) {
 		struct rf_rx_gain rx_gain = {0};
-		mutex_lock(&phy->indio_dev->mlock);
+		mutex_lock(&phy->lock);
 		ret = ad9361_get_rx_gain(phy, (entry->cmd == DBGFS_RXGAIN_1) ?
 				1 : 2, &rx_gain);
-		mutex_unlock(&phy->indio_dev->mlock);
+		mutex_unlock(&phy->lock);
 		if (ret < 0)
 			return ret;
 
@@ -8221,7 +8284,7 @@ static ssize_t ad9361_debugfs_write(struct file *file,
 	case DBGFS_INIT:
 		if (!(ret == 1 && val == 1))
 			return -EINVAL;
-		mutex_lock(&phy->indio_dev->mlock);
+		mutex_lock(&phy->lock);
 		clk_set_rate(phy->clks[TX_SAMPL_CLK], 1);
 		clk_set_parent(phy->clks[RX_RFPLL], phy->clk_ext_lo_rx);
 		clk_set_parent(phy->clks[TX_RFPLL], phy->clk_ext_lo_tx);
@@ -8230,7 +8293,7 @@ static ssize_t ad9361_debugfs_write(struct file *file,
 		ad9361_clks_disable(phy);
 		ad9361_clear_state(phy);
 		ret = ad9361_setup(phy);
-		mutex_unlock(&phy->indio_dev->mlock);
+		mutex_unlock(&phy->lock);
 		if (ret < 0)
 			return ret;
 
@@ -8238,9 +8301,9 @@ static ssize_t ad9361_debugfs_write(struct file *file,
 	case DBGFS_LOOPBACK:
 		if (ret != 1)
 			return -EINVAL;
-		mutex_lock(&phy->indio_dev->mlock);
+		mutex_lock(&phy->lock);
 		ret = ad9361_bist_loopback(phy, val);
-		mutex_unlock(&phy->indio_dev->mlock);
+		mutex_unlock(&phy->lock);
 		if (ret < 0)
 			return ret;
 
@@ -8249,9 +8312,9 @@ static ssize_t ad9361_debugfs_write(struct file *file,
 	case DBGFS_BIST_PRBS:
 		if (ret != 1)
 			return -EINVAL;
-		mutex_lock(&phy->indio_dev->mlock);
+		mutex_lock(&phy->lock);
 		ret = ad9361_bist_prbs(phy, val);
-		mutex_unlock(&phy->indio_dev->mlock);
+		mutex_unlock(&phy->lock);
 		if (ret < 0)
 			return ret;
 
@@ -8260,9 +8323,9 @@ static ssize_t ad9361_debugfs_write(struct file *file,
 	case DBGFS_BIST_TONE:
 		if (ret != 4)
 			return -EINVAL;
-		mutex_lock(&phy->indio_dev->mlock);
+		mutex_lock(&phy->lock);
 		ret = ad9361_bist_tone(phy, val, val2, val3, val4);
-		mutex_unlock(&phy->indio_dev->mlock);
+		mutex_unlock(&phy->lock);
 		if (ret < 0)
 			return ret;
 
@@ -8271,9 +8334,9 @@ static ssize_t ad9361_debugfs_write(struct file *file,
 	case DBGFS_MCS:
 		if (ret != 1)
 			return -EINVAL;
-		mutex_lock(&phy->indio_dev->mlock);
+		mutex_lock(&phy->lock);
 		ret = ad9361_mcs(phy, val);
-		mutex_unlock(&phy->indio_dev->mlock);
+		mutex_unlock(&phy->lock);
 		if (ret < 0)
 			return ret;
 
@@ -8285,10 +8348,10 @@ static ssize_t ad9361_debugfs_write(struct file *file,
 
 		if (phy->pdata->cal_sw1_gpio &&
 			phy->pdata->cal_sw2_gpio) {
-			mutex_lock(&phy->indio_dev->mlock);
+			mutex_lock(&phy->lock);
 			gpiod_set_value(phy->pdata->cal_sw1_gpio, !!(val & BIT(0)));
 			gpiod_set_value(phy->pdata->cal_sw2_gpio, !!(val & BIT(1)));
-			mutex_unlock(&phy->indio_dev->mlock);
+			mutex_unlock(&phy->lock);
 		} else {
 			return -ENODEV;
 		}
@@ -8298,9 +8361,9 @@ static ssize_t ad9361_debugfs_write(struct file *file,
 	case DBGFS_DIGITAL_TUNE:
 		if (ret != 2)
 			return -EINVAL;
-		mutex_lock(&phy->indio_dev->mlock);
+		mutex_lock(&phy->lock);
 		ret = ad9361_dig_tune(phy, val, val2);
-		mutex_unlock(&phy->indio_dev->mlock);
+		mutex_unlock(&phy->lock);
 		if (ret < 0)
 			return ret;
 
@@ -8337,7 +8400,7 @@ static ssize_t ad9361_debugfs_write(struct file *file,
 			return -EINVAL;
 		}
 
-		mutex_lock(&phy->indio_dev->mlock);
+		mutex_lock(&phy->lock);
 		ctrl->gpo_manual_mode_enable_mask &= ~mask;
 		ctrl->gpo_manual_mode_enable_mask |= val3;
 
@@ -8357,7 +8420,7 @@ static ssize_t ad9361_debugfs_write(struct file *file,
 		if (!(val3 & GPO_MANUAL_SELECT))
 			ad9361_spi_write(phy->spi, REG_EXTERNAL_LNA_CTRL,
 					 val3 | GPO_MANUAL_SELECT);
-		mutex_unlock(&phy->indio_dev->mlock);
+		mutex_unlock(&phy->lock);
 		if (ret < 0)
 			return ret;
 
@@ -8577,10 +8640,10 @@ static struct ad9361_phy_platform_data
 			   &pdata->tdd_skip_vco_cal);
 
 	ad9361_of_get_u32(iodev, np, "adi,tx-fastlock-delay-ns", 0,
-			  &pdata->rx_fastlock_delay_ns);
+			  &pdata->tx_fastlock_delay_ns);
 
 	ad9361_of_get_u32(iodev, np, "adi,rx-fastlock-delay-ns", 0,
-			  &pdata->tx_fastlock_delay_ns);
+			  &pdata->rx_fastlock_delay_ns);
 
 	ad9361_of_get_bool(iodev, np, "adi,rx-fastlock-pincontrol-enable",
 			   &pdata->trx_fastlock_pinctrl_en[0]);
@@ -8849,6 +8912,8 @@ static struct ad9361_phy_platform_data
 	ad9361_of_get_u32(iodev, np, "adi,fagc-lock-level-gain-increase-upper-limit", 5,
 			&pdata->gain_ctrl.f_agc_lock_level_gain_increase_upper_limit); /* 0x118 0..63 */
 		/* Fast AGC - Peak Detectors and Final Settling */
+  ad9361_of_get_bool(iodev, np, "adi,fagc-dig-sat-ovrg-enable",
+	    &pdata->gain_ctrl.f_agc_dig_sat_ovrg_en); /* 0x101:7 (full table) */
 	ad9361_of_get_u32(iodev, np, "adi,fagc-lpf-final-settling-steps", 1,
 			&pdata->gain_ctrl.f_agc_lpf_final_settling_steps); /* 0x112:6 0..3 (Post Lock Level Step)*/
 	ad9361_of_get_u32(iodev, np, "adi,fagc-lmt-final-settling-steps", 1,
@@ -9346,7 +9411,7 @@ ad9361_gt_bin_write(struct file *filp, struct kobject *kobj,
 	if (IS_ERR_OR_NULL(table))
 		return PTR_ERR(table);
 
-	mutex_lock(&phy->indio_dev->mlock);
+	mutex_lock(&phy->lock);
 	ad9361_free_gt(phy, phy->gt_info);
 
 	st->current_table = -1;
@@ -9356,7 +9421,7 @@ ad9361_gt_bin_write(struct file *filp, struct kobject *kobj,
 		clk_get_rate(phy->clks[RX_RFPLL])),
 		GT_RX1 + GT_RX2);
 
-	mutex_unlock(&phy->indio_dev->mlock);
+	mutex_unlock(&phy->lock);
 
 	return count;
 }
@@ -9430,6 +9495,16 @@ static int ad9361_spi_check(struct spi_device *spi)
 	return t.effective_speed_hz;
 }
 
+static void ad9361_remove_ext_band_control(void *phy)
+{
+	ad9361_unregister_ext_band_control(phy);
+}
+
+static void ad9361_clk_del_provider(void *of_node)
+{
+	of_clk_del_provider(of_node);
+}
+
 static int ad9361_probe(struct spi_device *spi)
 {
 	struct iio_dev *indio_dev;
@@ -9459,6 +9534,7 @@ static int ad9361_probe(struct spi_device *spi)
 	phy->indio_dev = indio_dev;
 	phy->spi = spi;
 	phy->clk_refin = clk;
+	mutex_init(&phy->lock);
 
 	ad9361_init_state(phy);
 
@@ -9493,6 +9569,10 @@ static int ad9361_probe(struct spi_device *spi)
 			 "%s: failed to initialize ext band control\n",
 			 __func__);
 
+	ret = devm_add_action_or_reset(&spi->dev, ad9361_remove_ext_band_control, phy);
+	if (ret)
+		return ret;
+
 	phy->gt_info = ad9361_adi_gt_info;
 
 	ad9361_request_gt(phy, NULL);
@@ -9500,14 +9580,8 @@ static int ad9361_probe(struct spi_device *spi)
 	ad9361_reset(phy);
 
 	hz = ad9361_spi_check(spi);
-	if (hz < 0) {
+	if (hz < 0)
 		return hz;
-	}
-
-	ret = ad9361_spi_read(spi, REG_PRODUCT_ID);
-	if (ret < 0) {
-		return ret;
-	}
 
 	rev = ret & REV_MASK;
 
@@ -9526,12 +9600,16 @@ static int ad9361_probe(struct spi_device *spi)
 
 	ret = ad9361_setup(phy);
 	if (ret < 0)
-		goto out_unregister_notifier;
+		return ret;
 
 	ret = of_clk_add_provider(spi->dev.of_node,
 			    of_clk_src_onecell_get, &phy->clk_data);
 	if (ret)
-		goto out_disable_clocks;
+		return ret;
+
+	ret = devm_add_action_or_reset(&spi->dev, ad9361_clk_del_provider, spi->dev.of_node);
+	if (ret)
+		return ret;
 
 	sysfs_bin_attr_init(&phy->bin);
 	phy->bin.attr.name = "filter_fir_config";
@@ -9560,19 +9638,18 @@ static int ad9361_probe(struct spi_device *spi)
 	indio_dev->num_channels = ARRAY_SIZE(ad9361_phy_chan) -
 		(phy->pdata->rx2tx2 ? 0 : 2);
 
-	ret = iio_device_register(indio_dev);
+	ret = devm_iio_device_register(&spi->dev, indio_dev);
 	if (ret < 0)
-		goto out_clk_del_provider;
+		return ret;
 	ret = ad9361_register_axi_converter(phy);
 	if (ret < 0)
-		goto out_iio_device_unregister;
+		return ret;
 	ret = sysfs_create_bin_file(&indio_dev->dev.kobj, &phy->bin);
 	if (ret < 0)
-		goto out_iio_device_unregister;
+		return ret;
 	ret = sysfs_create_bin_file(&indio_dev->dev.kobj, &phy->bin_gt);
 	if (ret < 0)
-		goto out_iio_device_unregister;
-
+		return ret;
 
 	ret = ad9361_register_debugfs(indio_dev);
 	if (ret < 0)
@@ -9586,34 +9663,6 @@ static int ad9361_probe(struct spi_device *spi)
 				__func__, rev);
 
 	return 0;
-
-out_iio_device_unregister:
-	iio_device_unregister(indio_dev);
-out_clk_del_provider:
-	of_clk_del_provider(spi->dev.of_node);
-out_disable_clocks:
-	ad9361_clks_disable(phy);
-out_unregister_notifier:
-	clk_notifier_unregister(phy->clks[RX_RFPLL], &phy->clk_nb_rx);
-	clk_notifier_unregister(phy->clks[TX_RFPLL], &phy->clk_nb_tx);
-
-	return ret;
-}
-
-static int ad9361_remove(struct spi_device *spi)
-{
-	struct ad9361_rf_phy *phy = ad9361_spi_to_phy(spi);
-
-	ad9361_unregister_ext_band_control(phy);
-	sysfs_remove_bin_file(&phy->indio_dev->dev.kobj, &phy->bin_gt);
-	sysfs_remove_bin_file(&phy->indio_dev->dev.kobj, &phy->bin);
-	iio_device_unregister(phy->indio_dev);
-	of_clk_del_provider(spi->dev.of_node);
-	clk_notifier_unregister(phy->clks[RX_RFPLL], &phy->clk_nb_rx);
-	clk_notifier_unregister(phy->clks[TX_RFPLL], &phy->clk_nb_tx);
-	ad9361_clks_disable(phy);
-
-	return 0;
 }
 
 static const struct spi_device_id ad9361_id[] = {
@@ -9625,23 +9674,11 @@ static const struct spi_device_id ad9361_id[] = {
 };
 MODULE_DEVICE_TABLE(spi, ad9361_id);
 
-static const struct of_device_id ad9361_of_match[] = {
-	{ .compatible = "adi,ad9361" },
-	{ .compatible = "adi,ad9364" },
-	{ .compatible = "adi,ad9361-2x" },
-	{ .compatible = "adi,ad9363a" },
-	{}
-};
-MODULE_DEVICE_TABLE(of, ad9361_of_match);
-
 static struct spi_driver ad9361_driver = {
 	.driver = {
-		.name		= "ad9361",
-		.owner		= THIS_MODULE,
-		.of_match_table	= ad9361_of_match,
+		.name	= "ad9361",
 	},
 	.probe		= ad9361_probe,
-	.remove		= ad9361_remove,
 	.id_table	= ad9361_id,
 };
 module_spi_driver(ad9361_driver);
