From 2d9e991c0e0405a51788649d5da89c8b5c14912c Mon Sep 17 00:00:00 2001
From: Petrolo Vincenzo <s295043@studenti.polito.it>
Date: Sun, 12 Oct 2025 13:44:04 +0200
Subject: [PATCH] NO CF_AXI_DDS

---
 drivers/sdr.c | 8 ++++++++
 drivers/sdr.h | 4 ++++
 2 files changed, 12 insertions(+)

diff --git a/drivers/sdr.c b/drivers/sdr.c
index f5e9144..a053769 100644
--- a/drivers/sdr.c
+++ b/drivers/sdr.c
@@ -50,7 +50,9 @@
 #include <../../drivers/iio/adc/ad9361_regs.h>
 #include <../../drivers/iio/adc/ad9361.h>
 #include <../../drivers/iio/adc/ad9361_private.h>
+#if IS_ENABLED(CONFIG_IIO_FREQUENCY_CF_AXI_DDS)
 #include <../../drivers/iio/frequency/cf_axi_dds.h>
+#endif
 
 #include "nl80211_testmode_def.h"
 #include "hw_def.h"
@@ -59,7 +61,9 @@
 
 #ifndef RFSoC4x2
 extern int ad9361_do_calib_run(struct ad9361_rf_phy *phy, u32 cal, int arg); 
+#if IS_ENABLED(CONFIG_IIO_FREQUENCY_CF_AXI_DDS)
 extern int cf_axi_dds_datasel(struct cf_axi_dds_state *st, int channel, enum dds_data_select sel);
+#endif
 extern struct ad9361_rf_phy* ad9361_spi_to_phy(struct spi_device *spi);
 extern int ad9361_tx_mute(struct ad9361_rf_phy *phy, u32 state);
 extern int ad9361_ctrl_outs_setup(struct ad9361_rf_phy *phy, struct ctrl_outs_control *ctrl);
@@ -68,7 +72,9 @@ extern int ad9361_spi_read(struct spi_device *spi, u32 reg);
 extern int ad9361_get_tx_atten(struct ad9361_rf_phy *phy, u32 tx_num);
 #else
 int ad9361_do_calib_run(struct ad9361_rf_phy *phy, u32 cal, int arg){return(0);}; 
+#if IS_ENABLED(CONFIG_IIO_FREQUENCY_CF_AXI_DDS)
 int cf_axi_dds_datasel(struct cf_axi_dds_state *st, int channel, enum dds_data_select sel){return(0);}; 
+#endif
 struct ad9361_rf_phy* ad9361_spi_to_phy(struct spi_device *spi){return(0);}; 
 int ad9361_tx_mute(struct ad9361_rf_phy *phy, u32 state){return(0);}; 
 int ad9361_ctrl_outs_setup(struct ad9361_rf_phy *phy, struct ctrl_outs_control *ctrl){return(0);}; 
@@ -2304,6 +2310,7 @@ static int openwifi_dev_probe(struct platform_device *pdev)
       goto err_free_dev;
     }
 
+    #if IS_ENABLED(CONFIG_IIO_CF_AXI_DDS)
     priv->dds_st = iio_priv(tmp_indio_dev);
     if (!(priv->dds_st)) {
       printk(KERN_ERR "%s iio_priv failed\n",sdr_compatible_str);
@@ -2313,6 +2320,7 @@ static int openwifi_dev_probe(struct platform_device *pdev)
     printk("%s openwifi_dev_probe: cf-ad9361-dds-core-lpc dds_st->version %08x chip_info->name %s\n",sdr_compatible_str,priv->dds_st->version,priv->dds_st->chip_info->name);
     cf_axi_dds_datasel(priv->dds_st, -1, DATA_SEL_DMA);
     printk("%s openwifi_dev_probe: cf_axi_dds_datasel DATA_SEL_DMA\n",sdr_compatible_str);
+    #endif
 
     // //-------------find driver: axi_ad9361 hdl ref design module, adc channel---------------
     // turn off radio by muting tx
diff --git a/drivers/sdr.h b/drivers/sdr.h
index 639a29f..7a0d557 100644
--- a/drivers/sdr.h
+++ b/drivers/sdr.h
@@ -347,6 +347,7 @@ static const u16 wifi_n_dbps_ht_table[16] =        {26, 26, 26,  26, 26, 52,  78
 // static const u8 wifi_mcs_table[8] =             {6,9,12,18,24,36,48,54};
 // static const u8 wifi_mcs_table_phy_tx[8]    =   {11,15,10,14,9,13,8,12};
 
+#if IS_ENABLED(CONFIG_IIO_CF_AXI_DDS)
 // ===== copy from adi-linux/drivers/iio/frequency/cf_axi_dds.c =====
 struct cf_axi_dds_state {
   struct device               *dev_spi;
@@ -375,6 +376,7 @@ struct cf_axi_dds_state {
   struct notifier_block       clk_nb;
 };
 // ===== end of copy from adi-linux/drivers/iio/frequency/cf_axi_dds.c =====
+#endif
 
 struct openwifi_stat {
   u32 stat_enable;
@@ -451,7 +453,9 @@ struct openwifi_priv {
   enum openwifi_hardware_type  hardware_type;
   enum openwifi_fpga_type      fpga_type;
 
+#if IS_ENABLED(CONFIG_IIO_CF_AXI_DDS)
   struct cf_axi_dds_state      *dds_st;  //axi_ad9361 hdl ref design module, dac channel
+#endif
   struct axiadc_state          *adc_st;      //axi_ad9361 hdl ref design module, adc channel
   struct ad9361_rf_phy         *ad9361_phy; //ad9361 chip
   struct ctrl_outs_control     ctrl_out;
-- 
2.51.0

