From 4502905d513f1aad0ab048d21bfb9e2e38295361 Mon Sep 17 00:00:00 2001
From: Petrolo Vincenzo <s295043@studenti.polito.it>
Date: Sun, 12 Oct 2025 14:18:31 +0200
Subject: [PATCH] fix

---
 drivers/sdrctl_intf.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/sdrctl_intf.c b/drivers/sdrctl_intf.c
index ef91be9..9184396 100644
--- a/drivers/sdrctl_intf.c
+++ b/drivers/sdrctl_intf.c
@@ -52,14 +52,14 @@ static int openwifi_testmode_cmd(struct ieee80211_hw *hw, struct ieee80211_vif *
     }
     return 0;
   case OPENWIFI_CMD_GET_SLICE_IDX:
-    skb = (struct sk_buff *)cfg80211_testmode_alloc_reply_skb(hw->wiphy, nla_total_size(sizeof(u32)));
+    skb = (struct sk_buff *)cfg80211_vendor_cmd_alloc_reply_skb(hw->wiphy, nla_total_size(sizeof(u32)));
     if (!skb)
       return -ENOMEM;
     tmp = priv->slice_idx;
     if (nla_put_u32(skb, OPENWIFI_ATTR_SLICE_IDX, tmp))
       goto nla_put_failure;
     printk("%s get openwifi slice_idx in hex: %08x\n", sdr_compatible_str, tmp);
-    return cfg80211_testmode_reply(skb);
+    return cfg80211_vendor_cmd_reply(skb);
   case OPENWIFI_CMD_SET_ADDR:
     if (!tb[OPENWIFI_ATTR_ADDR])
       return -EINVAL;
-- 
2.51.0

